components/merino_client/src/merino_client.udl

```diff
@@ -26,14 +26,19 @@ dictionary MerinoClientFetchOptions {
     sequence<string>? providers = null;
 };
 
-dictionary MerinoSuggestion {
+dictionary MerinoSuggestionDetails {
     string title;
     string url;
-    string provider;
     boolean is_sponsored;
     f64 score;
     string? icon;
-    string request_id;
+};
+
+[Enum]
+interface MerinoSuggestion {
```

Comment:
thought: Even though we can't have inheritance, I've found it super helpful to have accessors for the details in both Kotlin and Swift:

```kt
val MerinoSuggestion.details: MerinoSuggestionDetails
    get() = when (this) {
        is MerinoSuggestion.Adm -> details
        is MerinoSuggestion.TopPicks -> details
        is MerinoSuggestion.Other -> details
    }
```

```swift
extension MerinoSuggestion {
    var details: MerinoSuggestionDetails {
        switch self {
        case .adm(let details, _, _, _, _, _):
            return details
        case .topPicks(let details, _, _):
            return details
        case .other(let details, _):
            return details
        }
    }
}
```

That lets us write code like `suggestion.details.url` and `suggestion.details.icon`, instead of having to pattern-match on the suggestion with `when`/`switch`, and extract the details every time we want a base suggestion field.

UniFFI enum interfaces can't have methods or getters, so we can't put that accessor in the UDL—but we can still add [hand-written Kotlin](https://mozilla.github.io/application-services/book/howtos/adding-a-new-component.html#the-kotlin-bindings) and [Swift](https://mozilla.github.io/application-services/book/howtos/adding-a-new-component.html#creating-the-directory-structure) source files with those extensions to our component! The FxA client does something like that in its [Kotlin](https://github.com/mozilla/application-services/blob/f0cf5491ce6f3ccdd80913b9a9a4c3568c6cb2da/components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/Config.kt) and [Swift](https://github.com/mozilla/application-services/blob/f0cf5491ce6f3ccdd80913b9a9a4c3568c6cb2da/components/fxa-client/ios/FxAClient/FxAccountConfig.swift) bindings.

This would be a great exercise for getting more familiar with the component directory structure, and updating the Xcode project to include the hand-written Swift extension!

/cc @tiftran

Reply:
> thought: Even though we can't have inheritance, I've found it super helpful to have accessors for the details in both Kotlin and Swift:

That's really nice! I wonder if we can teach UniFFI about that?

> UniFFI enum interfaces can't have methods or getters, so we can't put that accessor in the UDL—but we can still add [hand-written Kotlin](https://mozilla.github.io/application-services/book/howtos/adding-a-new-component.html#the-kotlin-bindings) and [Swift](https://mozilla.github.io/application-services/book/howtos/adding-a-new-component.html#creating-the-directory-structure) source files with those extensions to our component!

This raises an interesting point. The hand-written foreign code next to the component was initially hugely valuable - our hand-written FFI was so painful that wrappers made a lot of sense. However, as UniFFI means the bindings are quite ergonomic, the app-services team has had an informal decision to keep these hand-wrappers as small as possible - if at all. Part of the reason was that we found understanding old code became quite difficult, particularly for Fenix - by the time our component ended up being consumed we were often many levels of indirection away from the component itself - we have our hand-wrappers, a couple of wrappers in android-components (often a "concept" and a "service"), then the final consumer. Nimbus has gone the other way (the "canonical" wrappers are next to their component). I don't think there is a single correct answer here, but it is a tension that we still feel.

(One example of what I'm referring to is our [autofill component](https://github.com/mozilla/application-services/tree/main/components/autofill) - the only [hand-written kotlin there is working around a UniFFI limitation](https://github.com/mozilla/application-services/blob/main/components/autofill/android/src/main/java/mozilla/appservices/autofill/CreateKey.kt) (or probably more accurately, a limitation in our megazord layout) - ie, so similar to your example above. But any additional ergonomic improvements tends to be in our consumer code (eg, in [android-components](https://github.com/mozilla-mobile/firefox-android/blob/main/android-components/components/service/sync-autofill/src/main/java/mozilla/components/service/sync/autofill/Types.kt))

Reply:
> That's really nice! I wonder if we can teach UniFFI about that?

I filed https://github.com/mozilla/uniffi-rs/issues/1470. 😄

> However, as UniFFI means the bindings are quite ergonomic, the app-services team has had an informal decision to keep these hand-wrappers as small as possible - if at all.

_nod_ That makes sense, for all the reasons you mentioned—having a substantial amount of hand-written code in Application Services adds another layer of wrapping that can hide important implementation details. I totally agree with your approach of keeping that layer as small as possible (if at all), and saving the ergonomics for our consumers.